modules/catalog/src/main/java/org/apache/ignite/internal/catalog/Catalog.java

```diff
@@ -155,4 +172,23 @@ public Collection<CatalogZoneDescriptor> zones() {
     public String toString() {
         return S.toString(this);
     }
+
+    private static Int2ObjectMap<List<CatalogIndexDescriptor>> toIndexesByTableId(Collection<CatalogSchemaDescriptor> schemas) {
+        Int2ObjectMap<List<CatalogIndexDescriptor>> indexesByTableId = new Int2ObjectOpenHashMap<>();
+
+        for (CatalogSchemaDescriptor schema : schemas) {
+            for (CatalogTableDescriptor table : schema.tables()) {
+                int tableId = table.id();
+
+                List<CatalogIndexDescriptor> tableIndexes = Arrays.stream(schema.indexes())
```

Comment:
We may have a substantial number of tables and indexes. Currently, for `N` tables and `M` indexes, your code looks like `O(N*M*foo(M))`. I struggle to estimate `foo` in my head (sorting overhead), but it's definitely below `log(M)`. The main problem is in `N*M` multiplication.

What if we only iterated over the `schema.indexes()` once, creating a list in `indexesByTableId` lazily and sorting everything at the very end? Complexity would become about `O((N+M)*foo(M))`, I believe. We reduce it be removing extra loop and filtering.
In all calculations I assumed hash-table access to be `O(1)`, I mention this just in case